src/main/scala/calc/Tree.scala

```diff
@@ -1,6 +1,30 @@
 package calc
 
-import org.scalajs.core.ir.Position
+import org.scalajs.core.ir.{Position, Types => irtpe}
+
+sealed trait Type {
+  def irtype: irtpe.Type
+}
+
+case class TAny(name: String) extends Type {
+  def irtype = irtpe.AnyType
+}
+
+case object TDouble extends Type {
+  def irtype = irtpe.DoubleType
+}
+
+case class TFun(arity: Int) extends Type {
+  def irtype = irtpe.AnyType
+}
+
+case class TStaticForeignFun(clsName: String, method: String, arity: Int) extends Type {
+  def irtype = irtpe.AnyType
+  override def equals(o: Any) = o match {
+    case that: TStaticForeignFun => this.arity == that.arity
+    case _ => false
+  }
```

Comment:
Yeah, ideally foreign functions information should be stored elsewhere, not encoded in the type.

This was implemented before my eta reduce so that I don't need to include environment as additional parameter in my compiler code.